packages/composite-checkout-wpcom/src/hooks/use-shopping-cart.ts

```diff
@@ -142,6 +172,22 @@ export function useShoppingCart(
 			const response = await setServerCart( prepareRequestCart( responseCart ) );
 			debug( 'cart sent; new responseCart is', response );
 			setResponseCart( response );
+
+			if ( couponStatus === 'pending' ) {
+				if ( response.is_coupon_applied ) {
+					showAddCouponSuccessMessage( response.coupon );
```

Comment:
I'm surprised this doesn't trigger a linter warning since `showAddCouponSuccessMessage` might change and it's not listed as a dependency of this `useEffect`. 🤔 

Also, since this is not related to the cart directly, might this make more sense as its own `useEffect`? That is, an effect to update the coupon status when the cart updates? One way I like to think about useEffect is: can the meaning of this block be easily summarized into a function call? So previously this maybe could become `useSendCartToServer`, but once you add the coupon stuff, which is really derived state, it becomes a little more muddy.

Reply:
> might this make more sense as its own useEffect?

I tried making this change the naive way (just moving the outer if block) but in testing it appears not to work, maybe because the order of effects matters.

I need to document this better but the intention is to model the coupon interaction using a state machine, with state transitions triggered by the submitCoupon callback and the update cart handler.